### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by rahmal**
- Re-factor app into MVC pattern
- Breakup `app.rb` into controllers
- Add authentication via Github
- Make authentication required (so we have a better papertrail of who does what)

### CODE REVIEW

**Comment by soonahn** [comment]
@rahmal is this still something we should be merging and working towards, or is this okay to close?

**Comment by rahmal** [comment]
Don't close it. I've been keeping it up to date and in sync with master. It
works locally but I just haven't found a good time to merge and deploy it.
I didn't want to take the chance on pushing it in the middle of the day
when honesty is busy, but I want it in.

Rahmal Conda / Technical Lead

Grand Rounds

Office: 1-800-929-0926  / Fax: 1-888-323-5997

360 3rd St Suite 425, San Francisco, CA 94107

http://www.grandrounds.com ​

[image: Facebook] [image: Twitter] <https://twitter.com/GrandRoundsInc> [image:
Linkedin] <https://www.linkedin.com/company/grandrounds>

**Comment by soonahn** [comment]
@rahmal I think @alexebird has set up a development area for the Honesty machines.  So you can run this branch without worrying about taking down what Honesty.

**Comment by rahmal** [comment]
Actually, I do remember that. Thanks for reminding me.

Rahmal Conda / Technical Lead

Grand Rounds

Office: 1-800-929-0926  / Fax: 1-888-323-5997

360 3rd St Suite 425, San Francisco, CA 94107

http://www.grandrounds.com ​

[image: Facebook] [image: Twitter] <https://twitter.com/GrandRoundsInc> [image:
Linkedin] <https://www.linkedin.com/company/grandrounds>

**Comment by kenberland** [comment]
Is this really 71 changed files?  Can we get a cleaner diff?

**Comment by rahmal** [comment]
I'll create a new PR for it.

Rahmal Conda / Technical Lead

Grand Rounds

Office: 1-800-929-0926  / Fax: 1-888-323-5997

360 3rd St Suite 425, San Francisco, CA 94107

http://www.grandrounds.com ​

[image: Facebook] [image: Twitter] <https://twitter.com/GrandRoundsInc> [image:
Linkedin] <https://www.linkedin.com/company/grandrounds>

**Comment by rahmal** [comment]
Actually a new PR doesn't do anything. Same as this. However most of the changes is code being broken up. 40 of the 71 files are new. At least 4 of the 31 files left were the files that got broken-up *(including `app.rb` and `init.rb`)*.

**Comment by kenberland** [comment]
You need to find someone who will give this the proper amount of attention, then.

File: lib/app/public/css/users.css [hunk 1]
```diff
@@ -0,0 +1,4585 @@
+@import url("//fonts.googleapis.com/css?family=Open+Sans:300,400,600,700&amp;lang=en");
+/*
+ *
+ *   INSPINIA - Responsive Admin Theme
```
Comment (OAGr) [inline lib/app/public/css/users.css]: Is this a vendor asset? Can it be included as such?

File: lib/app/public/css/animate.css [hunk 2]
```diff
@@ -0,0 +1,2848 @@
+@charset "UTF-8";
+
+/*!
+Animate.css - http://daneden.me/animate
```
Comment (OAGr) [inline lib/app/public/css/animate.css]: Vendor asset?

File: lib/users.rb [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  u = Integrity::User.new
+  u.name = 'Ken Berland'
+  u.email = 'ken@grandrounds.com'
+  u.username = 'kenberland'
+  u.title = 'Head of Engineering'
+  u.gh_uid = '681863'
+  u.admin = true
+  u.save
+end
+
```
Comment (OAGr) [inline lib/users.rb]: Should this be modified to include @beghbali 's title now? :)

File: lib/users.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  u.email = 'grace.li@grandrounds.com'
+  u.username = 'glacier'
+  u.title = 'Engineer'
+  u.gh_uid = '233731'
+  u.admin = false
+  u.save
+end
+
+# Interns
+if !Integrity::User.first(username: 'mhfung')
```
Comment (OAGr) [inline lib/users.rb]: Remove

File: lib/users.rb [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+if !Integrity::User.first(username: 'glacier')
+  u = Integrity::User.new
+  u.name = 'Grace Li'
+  u.email = 'grace.li@grandrounds.com'
+  u.username = 'glacier'
+  u.title = 'Engineer'
+  u.gh_uid = '233731'
+  u.admin = false
+  u.save
+end
```
Comment (OAGr) [inline lib/users.rb]: If we are going this way we need Jesse, Josh, Richard, Nancy

**Comment by rahmal** [comment]
Ozzie has said he'll take a look, and I'm going to ask Alex to take a look when he gets a chance as well. Both are somewhat familiar with. I showed them the code about a month ago. Back then they both approved and even deployed it. But they rolled it back because it hadn't passed through the test env yet. I've been keeping it up to date and fixed some bugs since then.

File: lib/app/controllers/pancakes.rb [hunk 6]
```diff
@@ -0,0 +1,256 @@
+module Integrity
+  module Pancakes
+    def self.included(base)
+      base.class_eval do
+
+        # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+        # pancakes # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+        # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
```
Comment (OAGr) [inline lib/app/controllers/pancakes.rb]: I imagine this comment can go away, given the file is pancakes.rb

**Comment by OAGr** [comment]
In general I really like this refactor.  It split up a bunch of huge files and really seems like it would make the codebase more understandable.  

The vast majority of the diff seems to be moving code around and adding vendor assets.  It's not obvious what code is really new in all of that.  Would it be possible for you to make a few github comments to reference the new additions, specifically the logic you would like the most attention on?

In hindsight it would have been really nice if this were split up into multiple PRs; re-organization split from new functionality.  That said, I understand that there's been issues with the deployment of this and that sometimes things happen that way.

**Comment by beghbali** [comment]
I'm just catching up here. What's the current status on this PR? mergeable, requires CR changes, should be closed?

**Comment by beghbali** [comment]
@rahmal is this pr still valid?

**Comment by rahmal** [comment]
No, master has changed so much, I'll have to redo the merge. I'll create another PR when I've done it.

**Pull request closed without merging**